src/DsaBottomUp.cc

```diff
@@ -290,56 +369,126 @@ bool BottomUpAnalysis::runOnModule(Module &M, GraphMap &graphs) {
-                                                       *sm, do_sanity_checks);
-          if (!res) {
-            // continue;
-            llvm_unreachable("Simulation mapping check failed");
-          }
-          m_callee_caller_map.insert(
-              std::make_pair(dsaCS.getInstruction(), sm));
-
-          if (do_sanity_checks) {
-            // Check the simulation map is a function
-            if (!sm->isFunction())
-              errs() << "ERROR: simulation map for " << *dsaCS.getInstruction()
-                     << " is not a function!\n";
-            // Check that all nodes in the callee are mapped to one
-            // node in the caller graph
-            checkAllNodesAreMapped(*callee, calleeG, *sm);
-          }
-        }
-      }
+      /////
+      /// FIXME: Commenting this code breaks the SAS'17 code.
+      /////
+      // if (m_computeSimMap) {
+      //   // -- store the simulation maps from the SCC
+      //   for (auto &callRecord : *cgn) {
+      //     ImmutableCallSite CS(callRecord.first);
+      //     DsaCallSite dsaCS(CS);
+      //     const Function *callee = dsaCS.getCallee();
+      //     if (!callee || callee->isDeclaration() || callee->empty())
+      //       continue;
+
+      //     assert(graphs.count(dsaCS.getCaller()) > 0);
+      //     assert(graphs.count(dsaCS.getCallee()) > 0);
+
+      //     Graph &callerG = *(graphs.find(dsaCS.getCaller())->second);
+      //     Graph &calleeG = *(graphs.find(dsaCS.getCallee())->second);
+
+      //     SimulationMapperRef sm(new SimulationMapper());
+      //     bool res = Graph::computeCalleeCallerMapping(dsaCS, calleeG,
+      //     callerG,
+      //                                                  *sm,
+      //                                                  do_sanity_checks);
+      //     if (!res) {
+      //       // continue;
+      //       llvm_unreachable("Simulation mapping check failed");
+      //     }
+      //     m_callee_caller_map.insert(
+      //         std::make_pair(dsaCS.getInstruction(), sm));
+
+      //     if (do_sanity_checks) {
+      //       // Check the simulation map is a function
+      //       if (!sm->isFunction())
+      //         errs() << "ERROR: simulation map for " <<
+      //         *dsaCS.getInstruction()
+      //                << " is not a function!\n";
+      //       // Check that all nodes in the callee are mapped to one
+      //       // node in the caller graph
+      //       checkAllNodesAreMapped(*callee, calleeG, *sm);
+      //     }
+      //   }
+      // }
     }
 
     if (fGraph)
       fGraph->compress();
   }
 
+  /// Compute address-taken functions
+  /// FIXME: improve performance by maybe doing it together with the
+  ///        above loop that inlines graphs.
+  ///
+  /// TODO: this set should be refined by removing functions that have
+  ///       been resolved.
+  std::set<const Function *> addressTakenFunctions;
```

Comment:
You can query `llvm::Function` directly to whether its address is taken... Maybe this needs a better comment

Reply:
The set of address-taken functions changes over the bottom-up iterations. For the first iteration, yes we can ask `llvm::Function`. But after the first iteration, since we resolve callsites some functions' addresses might not be taken anymore

Reply:
This should be commented